File: formcraft-plm/src/main/java/fr/formcraft/model/entity/Product.java
package fr.formcraft.model.entity;

import fr.formcraft.model.enums.ProductState;
import fr.formcraft.model.enums.ProductType;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.UpdateTimestamp;

import java.math.BigDecimal;
import java.time.LocalDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Core product entity — represents any node in the PLM system:
 * finished products, semi-finished, raw materials, or packaging.
 * Mirrors beCPG's entity/node concept.
 */
@Entity
@Table(name = "products")
@Getter
@Setter
@NoArgsConstructor
public class Product {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false, unique = true, length = 50)
    private String code;

    @Column(nullable = false)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Enumerated(EnumType.STRING)
    @Column(name = "product_type", nullable = false, length = 30)
    private ProductType productType;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 30)
    private ProductState state = ProductState.DRAFT;

    @Column(length = 20)
    private String unit;

    @Column(name = "cost_per_kg", precision = 12, scale = 4)
    private BigDecimal costPerKg;

    /**
     * SPEL-based formula expression evaluated at formulation time.
     * Example: "protein * 4 + fat * 9 + carbohydrates * 4"
     * Variables are populated from the computed nutrient values.
     */
    @Column(name = "formula_expression", columnDefinition = "TEXT")
    private String formulaExpression;

    /**
     * Comma-separated allergen flags for quick lookup.
     * Example: "GLUTEN,EGGS,MILK"
     */
    @Column(name = "allergen_flags")
    private String allergenFlags;

    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt = LocalDateTime.now();

    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

    @Column(name = "created_by", length = 100)
    private String createdBy;

    @Column(name = "updated_by", length = 100)
    private String updatedBy;

    @Version
    private Long version;

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    @OrderBy("position ASC")
    private List<CompositionLine> compositionLines = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<NutrientValue> nutrientValues = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<FormulationResult> formulationResults = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<WorkflowTask> workflowTasks = new ArrayList<>();

    @OneToMany(mappedBy = "product", cascade = CascadeType.ALL, orphanRemoval = true, fetch = FetchType.LAZY)
    private List<QualityCheck> qualityChecks = new ArrayList<>();

    public boolean hasAllergen(String allergen) {
        if (allergenFlags == null || allergenFlags.isBlank()) {
            return false;
        }
        return allergenFlags.contains(allergen);
    }

    public boolean isFinishedProduct() {
        return ProductType.FINISHED_PRODUCT == productType;
    }

    public boolean isRawMaterial() {
        return ProductType.RAW_MATERIAL == productType;
    }
}

File: app/models/product.py
from __future__ import annotations

import datetime
from decimal import Decimal
from typing import List

from sqlalchemy import BigInteger, Column, Numeric, String, Text, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.enums.product_state import ProductState
from app.enums.product_type import ProductType


class Product(Base):
    __tablename__ = "products"

    id: Mapped[int] = mapped_column(BigInteger, primary_key=True, autoincrement=True)
    code: Mapped[str] = mapped_column(String(50), unique=True, nullable=False)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text)
    product_type: Mapped[str] = mapped_column(String(30), nullable=False)
    state: Mapped[str] = mapped_column(String(30), nullable=False, default=ProductState.DRAFT.value)
    unit: Mapped[str | None] = mapped_column(String(20))
    cost_per_kg: Mapped[Decimal | None] = mapped_column(Numeric(12, 4))
    formula_expression: Mapped[str | None] = mapped_column(Text)
    allergen_flags: Mapped[str | None] = mapped_column(Text)
    created_at: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False, default=datetime.datetime.utcnow)
    updated_at: Mapped[datetime.datetime] = mapped_column(DateTime, nullable=False, default=datetime.datetime.utcnow, onupdate=datetime.datetime.utcnow)
    created_by: Mapped[str | None] = mapped_column(String(100))
    updated_by: Mapped[str | None] = mapped_column(String(100))
    version: Mapped[int | None] = mapped_column(BigInteger, default=0)

    __mapper_args__ = {"version_id_col": version}

    composition_lines: Mapped[List["CompositionLine"]] = relationship(
        "CompositionLine", back_populates="product", cascade="all, delete-orphan",
        order_by="CompositionLine.position", foreign_keys="CompositionLine.product_id",
    )
    nutrient_values: Mapped[List["NutrientValue"]] = relationship(
        "NutrientValue", back_populates="product", cascade="all, delete-orphan",
    )
    formulation_results: Mapped[List["FormulationResult"]] = relationship(
        "FormulationResult", back_populates="product", cascade="all, delete-orphan",
    )
    workflow_tasks: Mapped[List["WorkflowTask"]] = relationship(
        "WorkflowTask", back_populates="product", cascade="all, delete-orphan",
    )
    quality_checks: Mapped[List["QualityCheck"]] = relationship(
        "QualityCheck", back_populates="product", cascade="all, delete-orphan",
    )

    def has_allergen(self, allergen: str) -> bool:
        if not self.allergen_flags:
            return False
        return allergen in self.allergen_flags.split(",")

    @property
    def is_finished_product(self) -> bool:
        return self.product_type == ProductType.FINISHED_PRODUCT.value

    @property
    def is_raw_material(self) -> bool:
        return self.product_type == ProductType.RAW_MATERIAL.value
